gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -314,6 +344,33 @@ public void handleApplicationReportArrivalEvent(ApplicationReportArrivalEvent ap
         LOGGER.error("Failed to close the " + GobblinYarnAppLauncher.class.getSimpleName(), ioe);
       } catch (TimeoutException te) {
         LOGGER.error("Timeout in stopping the service manager", te);
+      } finally {
+        if (this.emailNotificationOnShutdown) {
+          sendEmailOnShutdown(Optional.of(applicationReport));
+        }
+      }
+    }
+  }
+
+  @Subscribe
+  public void handleGetApplicationReportFailureEvent(
```

Comment:
No, this method should not be called concurrently. I just feel an `AtomicInteger` is a good choice for a counter although it's not really used for concurrency.